File: internal/store/memory/tasks.go
package memory

import (
	"context"
	"fmt"
	"slices"
	"time"

	"github.com/oklog/ulid/v2"
	"github.com/rakunlabs/at/internal/service"
	"github.com/rakunlabs/query"
)

func (m *Memory) ListTasks(_ context.Context, q *query.Query) (*service.ListResult[service.Task], error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	result := make([]service.Task, 0, len(m.tasks))
	for _, t := range m.tasks {
		result = append(result, t)
	}

	slices.SortFunc(result, func(a, b service.Task) int {
		if a.Title < b.Title {
			return -1
		}
		if a.Title > b.Title {
			return 1
		}
		return 0
	})

	return paginate(result, q), nil
}

func (m *Memory) GetTask(_ context.Context, id string) (*service.Task, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	t, ok := m.tasks[id]
	if !ok {
		return nil, nil
	}

	return &t, nil
}

func (m *Memory) CreateTask(_ context.Context, task service.Task) (*service.Task, error) {
	id := ulid.Make().String()
	now := time.Now().UTC().Format(time.RFC3339)

	rec := service.Task{
		ID:              id,
		OrganizationID:  task.OrganizationID,
		ProjectID:       task.ProjectID,
		GoalID:          task.GoalID,
		ParentID:        task.ParentID,
		AssignedAgentID: task.AssignedAgentID,
		Identifier:      task.Identifier,
		Title:           task.Title,
		Description:     task.Description,
		Status:          task.Status,
		PriorityLevel:   task.PriorityLevel,
		Priority:        task.Priority,
		Result:          task.Result,
		BillingCode:     task.BillingCode,
		RequestDepth:    task.RequestDepth,
		CreatedAt:       now,
		UpdatedAt:       now,
		CreatedBy:       task.CreatedBy,
		UpdatedBy:       task.UpdatedBy,
	}

	m.mu.Lock()
	m.tasks[id] = rec
	m.mu.Unlock()

	return &rec, nil
}

func (m *Memory) UpdateTask(_ context.Context, id string, task service.Task) (*service.Task, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.tasks[id]
	if !ok {
		return nil, nil
	}

	existing.OrganizationID = task.OrganizationID
	existing.ProjectID = task.ProjectID
	existing.GoalID = task.GoalID
	existing.ParentID = task.ParentID
	existing.AssignedAgentID = task.AssignedAgentID
	existing.Identifier = task.Identifier
	existing.Title = task.Title
	existing.Description = task.Description
	existing.Status = task.Status
	existing.PriorityLevel = task.PriorityLevel
	existing.Priority = task.Priority
	existing.Result = task.Result
	existing.BillingCode = task.BillingCode
	existing.RequestDepth = task.RequestDepth
	existing.StartedAt = task.StartedAt
	existing.CompletedAt = task.CompletedAt
	existing.CancelledAt = task.CancelledAt
	existing.HiddenAt = task.HiddenAt
	existing.UpdatedAt = now
	existing.UpdatedBy = task.UpdatedBy
	m.tasks[id] = existing

	return &existing, nil
}

func (m *Memory) DeleteTask(_ context.Context, id string) error {
	m.mu.Lock()
	delete(m.tasks, id)
	m.mu.Unlock()

	return nil
}

func (m *Memory) ListTasksByAgent(_ context.Context, agentID string) ([]service.Task, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	var result []service.Task
	for _, t := range m.tasks {
		if t.AssignedAgentID == agentID {
			result = append(result, t)
		}
	}

	slices.SortFunc(result, func(a, b service.Task) int {
		if a.Title < b.Title {
			return -1
		}
		if a.Title > b.Title {
			return 1
		}
		return 0
	})

	return result, nil
}

func (m *Memory) ListTasksByGoal(_ context.Context, goalID string) ([]service.Task, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	var result []service.Task
	for _, t := range m.tasks {
		if t.GoalID == goalID {
			result = append(result, t)
		}
	}

	slices.SortFunc(result, func(a, b service.Task) int {
		if a.Title < b.Title {
			return -1
		}
		if a.Title > b.Title {
			return 1
		}
		return 0
	})

	return result, nil
}

func (m *Memory) CheckoutTask(_ context.Context, taskID, agentID string) error {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.tasks[taskID]
	if !ok {
		return nil
	}

	if existing.CheckedOutBy != "" && existing.CheckedOutBy != agentID {
		return fmt.Errorf("task %q is already checked out by agent %q", taskID, existing.CheckedOutBy)
	}

	existing.CheckedOutBy = agentID
	existing.CheckedOutAt = now
	m.tasks[taskID] = existing

	return nil
}

func (m *Memory) ReleaseTask(_ context.Context, taskID string) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.tasks[taskID]
	if !ok {
		return nil
	}

	existing.CheckedOutBy = ""
	existing.CheckedOutAt = ""
	m.tasks[taskID] = existing

	return nil
}

File: _ui/src/lib/api/tasks.ts
import axios from 'axios';
import type { ListResult, ListParams } from './types';

const api = axios.create({ baseURL: 'api/v1' });

export const TASK_STATUSES = [
  'backlog', 'todo', 'in_progress', 'in_review', 'blocked', 'done', 'cancelled',
] as const;

export const TASK_STATUS_LABELS: Record<string, string> = {
  backlog: 'Backlog',
  open: 'Open',
  todo: 'To Do',
  in_progress: 'In Progress',
  in_review: 'In Review',
  blocked: 'Blocked',
  review: 'Review',
  completed: 'Completed',
  done: 'Done',
  cancelled: 'Cancelled',
};

export const TASK_PRIORITIES = ['critical', 'high', 'medium', 'low'] as const;

export const TASK_PRIORITY_LABELS: Record<string, string> = {
  critical: 'Critical',
  high: 'High',
  medium: 'Medium',
  low: 'Low',
};

export interface Task {
  id: string;
  organization_id: string;
  project_id: string;
  goal_id: string;
  parent_id: string;
  assigned_agent_id: string;
  identifier: string;
  title: string;
  description: string;
  status: string;
  priority_level: string;
  priority: number;
  result: string;
  billing_code: string;
  request_depth: number;
  // Per-task max iterations override. 0 = use the agent's default.
  // The iteration counter always starts fresh at 0 for each task run.
  max_iterations?: number;
  checked_out_by: string;
  checked_out_at: string;
  started_at: string;
  completed_at: string;
  cancelled_at: string;
  hidden_at: string;
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

export interface TaskWithSubtasks extends Task {
  sub_tasks?: TaskWithSubtasks[];
}

export async function listTasks(params?: ListParams): Promise<ListResult<Task>> {
  const res = await api.get<ListResult<Task>>('/tasks', { params });
  return res.data;
}

export async function getTask(id: string): Promise<Task> {
  const res = await api.get<Task>(`/tasks/${id}`);
  return res.data;
}

export async function getTaskWithSubtasks(id: string): Promise<TaskWithSubtasks> {
  const res = await api.get<TaskWithSubtasks>(`/tasks/${id}`, {
    params: { include: 'subtasks' },
  });
  return res.data;
}

export async function createTask(data: Partial<Task>): Promise<Task> {
  const res = await api.post<Task>('/tasks', data);
  return res.data;
}

export async function updateTask(id: string, data: Partial<Task>): Promise<Task> {
  const res = await api.put<Task>(`/tasks/${id}`, data);
  return res.data;
}

export async function deleteTask(id: string): Promise<void> {
  await api.delete(`/tasks/${id}`);
}

export async function checkoutTask(id: string, agentId: string): Promise<void> {
  await api.post(`/tasks/${id}/checkout`, { agent_id: agentId });
}

export async function releaseTask(id: string): Promise<void> {
  await api.post(`/tasks/${id}/release`);
}

export async function processTask(id: string): Promise<{ id: string; status: string }> {
  const res = await api.post<{ id: string; status: string }>(`/tasks/${id}/process`);
  return res.data;
}

export async function createTaskChat(id: string): Promise<{ id: string; agent_id: string; task_id: string; organization_id: string; name: string }> {
  const res = await api.post(`/tasks/${id}/chat`);
  return res.data;
}

export async function cancelTaskDelegation(id: string): Promise<{ message: string; task_id: string }> {
  const res = await api.post<{ message: string; task_id: string }>(`/tasks/${id}/cancel`);
  return res.data;
}

export interface ActiveDelegation {
  task_id: string;
  agent_id: string;
  org_id: string;
  started_at: string;
  duration: string;
}

export async function listActiveDelegations(): Promise<{ delegations: ActiveDelegation[] }> {
  const res = await api.get<{ delegations: ActiveDelegation[] }>('/active-delegations');
  return res.data;
}

export async function listTasksByAgent(agentId: string, params?: ListParams): Promise<ListResult<Task>> {
  const res = await api.get<ListResult<Task>>(`/agents/${agentId}/tasks`, { params });
  return res.data;
}
